Added implementation files and appropriate README

amrtanair opened the pull request:
This project is inspired by ["Minimally supervised induction of grammatical gender"](https://dl.acm.org/doi/10.3115/1073445.1073451) - Silviu Cucerzan, David Yarowsky.
This program uses a set of steps to compute the gender of a noun using a set of rules and morphology-based categorization.

The README has instructions to run the program.

amrtanair committed:
Added  code for predicting gender of german nouns

amrtanair committed:
add files and small edit to app

amrtanair committed:
Add a good README with images

jirka-x1 commented on app.py:
```diff
@@ -0,0 +1,73 @@
+
+	return suffixes_masc, suffixes_fem, suffixes_neut
+
+def calculate_gender(noun, suffixes_masc, suffixes_fem, suffixes_neut):
```
Use typehints (5.2 https://ufal.mff.cuni.cz/courses/npfl096-2014-su/project).

Here it should be probably 
```suggestion
def calculate_gender(noun: str, suffixes_masc: List[str], suffixes_fem: List[str], suffixes_neut: List[str]) -> str:
```

amrtanair replied:
fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+		    	gender = None
+
+		    	# german grammar rules for pronoun/nounsS
+	    		if pronoun in ["der", "die", "das", "dem", "des", "den"]:
```
it is not a pronoun but an article

amrtanair replied:
agreed and fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+import os.path
+
+# create intital gender-noun mapping from seed
+def initial_seeds():
```
typehints (return value) here and for all functions

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+	text = []
+	with open('files/train.txt') as train:
+	    for line in train:
+	    	text.append(nlp(line))
```
Simpler:
```suggestion
	with open('files/train.txt') as train:
	    	text = [nlp(line) for line in train]
```

amrtanair replied:
agreed and fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+	    pickle.dump(text, handle)
+
+def load_corpus():
+	obj = open('files/corpus.pkl', 'rb')
```
use `with`

amrtanair replied:
fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+	    for idx, token in enumerate(line):
+	    	if token.pos_ == "NOUN":
+	    		# get pronoun for current token
+		    	pronoun = (line[idx - 1]).text
```
call it a previous word, it can be an adjective, etc.

amrtanair replied:
agreed and fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+	    		# get pronoun for current token
+		    	pronoun = (line[idx - 1]).text
+		    	# get case of current token
+		    	case = token.morph.get("Case")
```
Well but this is "cheating". Spacy uses info about gender to recognize the case. Limit yourself to nonambiguous articles (das, des, dem)

amrtanair replied:
I agree with your point that it is "cheating". I was adding nouns to the seed only if they were verified to be correct. now, I am adding regardless of whether they are correctly classified or not. 

I do not know German and I followed this 
<img width="591" alt="image" src="https://user-images.githubusercontent.com/16919437/189951271-1e382dd4-2fe9-4c99-9af3-9066c6bf869e.png">
source: [German Definite Articles ](https://www.clozemaster.com/blog/german-definite-articles/) to choose articles that are unambiguous. I also added indefinite articles that seem unambiguous. The linked page mentions that the genitive case is not used much and hence I dropped it. Do let me know if I can change anything else

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+		    				gender = "Masc"
+		    			elif case == "Dat" or case == "Gen":
+		    				gender = "Fem"
+		    		elif pronoun == "die":
```
how do you know it is not plural? (nominative plural, _die_ is used for all genders)

amrtanair replied:
I am not using the article "die" anymore

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+			if s in suffixes_masc.keys():
+				suffixes_masc[s] = suffixes_masc[s] + 1
+			else:
+				suffixes_masc[s] = 1
```
use Counter (https://docs.python.org/3/library/collections.html#collections.Counter)
```suggestion
		suffixes_masc.update(suffixes)
```

amrtanair replied:
agreed and fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+		if suffixes_masc[key] < 3:
+			del suffixes_masc[key]
+		else:
+			c = c + suffixes_masc[key]
```
If possible, don't do two things in such a simple iteration.  Also use a better variable name than `c`
```suggestion
suffixes_masc = Counter({k: v for k, v in suffixes_masc.items() if c >= 3})
c = suffixes_masc.total()
```

amrtanair replied:
fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+			c = c + suffixes_masc[key]
+
+	for key, val in suffixes_masc.items():
+		suffixes_masc[key] = suffixes_masc[key]/c
```
```suggestion
	suffixes_masc = {key: val/c for key, val in suffixes_masc.items() }
```
(btw why do you write `suffixes_masc[key]/c` when you have `val` so `val/c` is enough)

amrtanair replied:
agreed and fixed. using list comprehensions and counter to calculate

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+	for key, val in suffixes_masc.items():
+		suffixes_masc[key] = suffixes_masc[key]/c
+
+	# Feminine
```
You have here 3x nearly the same code. Write a function for this (following my suggestions above). Call the function 3x with the gender as a parameter

amrtanair replied:
agreed and fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+	    		# probability of word being neutral/masculine/feminine
+		    	neut_p = 0
+		    	masc_p = 0
+		    	fem_p = 0
```
These should be floats (0.0), not integers (0)

amrtanair replied:
fixed

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+
+		    	# check if gender was calculated and spacy assigned gender exists. if it does and the assigned gender is corrected compared
+		    	# to the spacy assignmend, add it to the seed.
+		    	if gender and spacy_gender:
```
I am not sure I understand. You use spacy to expand you seed? But spacy is like gold data. So you should not consult them during calculating the gender.

amrtanair replied:
agreed and fixed. the explanation is given in the comment below https://github.com/amrtanair/german_noun_gender_predictor-/pull/1#issuecomment-1245620600

jirka-x1 commented on create_db.py:
```diff
@@ -0,0 +1,265 @@
+	return suffixes_masc, suffixes_fem, suffixes_neut
+
+def morphology_based_classification(corpus, seed, suffixes_masc, suffixes_fem, suffixes_neut):
+	# total number of nouns for which there was a calculated gender and a Spacy assigned gender
```
Separate the actual classification and evaluation. This way it is hard to see where is the boundary.

jirka-x1 replied:
I see you did not do this. I will accept this for the purpose of the class, but:

In real world, you do classification and evaluation separately. Imagine, you are running this as a product for some company 
- the classification is what your customers pay you for (they send you a noun and you return the gender) and 
- evaluation is what you run internally so that you can trust your product API (alongside of unit test, integration tests, etc)

amrtanair replied:
I think I have implemented this, In terms of classification, `create_db.py` is the evaluation file, it creates the "suffix bank" which is a collection of suffixes for every gender. This collection is then written to a file after pickling. It also calculates how well the suffix bank works with respect to the nouns already present in the corpus which is where the accuracy scores come in. 

In terms of evaluation, `app.py` is the file that actually uses the suffix bank(and the suffix pickle files for every gender)  and predicts the gender of a noun when given as input. It also has an API call to a german gender noun API to compare. All we need to provide a potential customer would be the three gender `.pkl` files and `app.py`. 

Is this what you are looking for?

amrtanair committed:
Fix #1 removed the spacy dependency

amrtanair commented:
Fix #1 involves:
1. Changed pronoun to articles
2. removed the condition where I would add the nouns to the seed only if they were correctly categorised. now, the nouns are added to the seed as they are classified. 
3. removed the dependency on spacy, where Spacy would provide the case of the noun. the case of the noun combined with the article used for it would give the gender, but now since the case has been removed, only the articles that are unambiguous are used. this has increased the accuracy for step 2 to 91% from an earlier ~70%. But, this also means that the accuracy after Step 3 has dropped to ~50% from an initial 60%. 
<img width="1382" alt="image" src="https://user-images.githubusercontent.com/16919437/189950468-518cd76d-d66b-4469-963e-5630e135c5fd.png">

5. The probabilities were initialised to 0, now they are initialised to 0.0
6. Changed the way the pickle object is loaded. 

I will soon add another fix that will address the rest of the issues.

amrtanair committed:
added counter, cleaned up code

amrtanair committed:
added type hinting and other fixes

amrtanair commented:
All the other fixes have been added. I was not sure of two changes:
1. The type hint for returning pickle objects. I have assigned 'Any' as the type in these cases
2. I was not sure on how to separate evaluation and classification so I did not change it at all. I could add another function for evaluation but i am not sure if I should
